python/terminal.py

```diff
@@ -0,0 +1,263 @@
+        the return/enter key. In a real-time interactive application we instead
+        want to receive each keystroke as it happens.
+
+        This context manager achieves that by setting 'cbreak' mode on the
+        the output tty stream. cbreak is a mode inbetween 'cooked mode', where
+        all the user's input is preprocessed, and 'raw mode' where none of it
+        is. Basically, in cbreak mode input like :kbd:`Control-c` will still
+        interrupt (i.e. 'break') the process, hence the name. Wikipedia is your
+        friend on this one!
+
+        :meth:`Root.run` uses this context manager for you to make your
+        application work in the correct way.
+        '''
+        if self.is_a_tty:
+            orig_tty_attrs = termios.tcgetattr(self.stream)
+            tty.setcbreak(self.stream)
+            try:
+                yield
+            finally:
+                termios.tcsetattr(
+                    self.stream, termios.TCSADRAIN, orig_tty_attrs)
+        else:
+            yield
+
+    @contextmanager
+    def nonblocking_input(self):
+        '''Context manager to set the :class:`Terminal`'s input file to read in
+        a non-blocking way.
+
+        Normally, reading from :attr:`sys.stdin` blocks, which is bad if we
+        want an interactive application that can also update information on the
+        screen without any input from the user! Therefore, we need to make
+        :meth:`Terminal.infile.read` a non-blocking operation, which just
+        returns nothing if we have no input.
+
+        :meth:`Root.run` uses this context manager for you to make your
+        application work in the correct way.
+        '''
+        # FIXME: do we handle restoring this during SIGTSTP?
+        if hasattr(self.infile, 'fileno'):
+            # Use fcntl to set stdin to non-blocking. WARNING - this is not
+            # particularly portable!
+            flags = fcntl.fcntl(self.infile, fcntl.F_GETFL)
+            fcntl.fcntl(self.infile, fcntl.F_SETFL, flags | os.O_NONBLOCK)
+            try:
+                yield
+            finally:
+                fcntl.fcntl(self.infile, fcntl.F_SETFL, flags)
+        else:
+            yield
+
+
+class LinePrintingTerminal(Terminal):
+    '''Allows multiple lines to be printed to the terminal and then overwritten
+    simply with another set of lines.
+    '''
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._lines_drawn = 0
+
+    def print_lines(self, lines):
+        for _ in range(self._lines_drawn - 1):
+            self.stream.write(self.move_up)
+        # Tmux bug? means move_x could be '' in error:
+        self.stream.write(self.move_x(1) or '\x1b[1G')
+        self.stream.write('\n'.join(lines))
+        self._lines_drawn = len(lines)
+
+
+class Keyboard:
+    '''Utility class for turning key escape sequences into human-parsable key
+    names.
+    '''
+    def __init__(self):
+        self._sequence_to_name = {
```

Comment:
Yeah, but we only make one instance of it. Could do a bunch of class level stuff. However, not worth refactoring in this area: if we're going to change anything, we might as well use the new `blessed` package, which seems to handle all that jazz. Biggest problem with that might be a few of the edge cases I've worked around, like suspending.